### PR TITLE
Implement golden-file matching for `integration_test` on Android and iOS devices

### PR DESCRIPTION
**Pull request by matanlurey**
Work towards https://github.com/flutter/flutter/issues/143299.
Work towards https://github.com/flutter/flutter/issues/160043.

---

This PR implements, end-to-end, support for `matchesGoldenFile` when (a) running with `package:integration_test` (b) on a device, such as an Android emulator, Android device, iOS simulator, or iOS device, where the _runner_ of a test file does not have process and local-file system access.

There are multiple parts to this PR; I could make it smaller than 1K lines, but the bulk of that is tests, and it would mean landing PRs that are incomplete and unused, which does not seem useful - so instead here is a quick overview of the PR's contents - questions/feedback welcome, and I am willing to break code out or land incremental refactors if requested.

1. Augmented `flutter_platform.dart` (used for iOS and Android), similar to [`flutter_web_platform.dart`](https://github.com/flutter/flutter/blob/1398dc7eecb696d302e4edb19ad79901e615ed56/packages/flutter_tools/lib/src/test/flutter_web_platform.dart#L117-L128), now creates and uses [`test_golden_comparator.dart`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/test/test_golden_comparator.dart) to proxy calls (coming from the VM service protocol) for golden-file updates and comparisons to a `flutter_tester` process. A full explanation of how (or why) it works this way is too hard to include here, but see https://github.com/flutter/flutter/pull/160215 for more details.
1. Added `VmServiceProxyGoldenFileComparator`, which is a currently unused (outside of a single e2e test) comparator that forwards calls to `compare` and `update` to the VM service protocol (of which, the other side of this is implemented above, in `flutter_platform.dart`. The idea is that this comparator would be used automatically when running in an integration test on a device that requires it (similar to how web works today), but that is **not** wired up yet and requires additional work in `flutter_tools`.
1. Added two unit tests (of both the client and server), and a full e2e-test using it to run `matchesGoldenFile`.

### CODE REVIEW

**Review by andrewkolos — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 1]
```diff
@@ -481,15 +510,64 @@ class FlutterPlatform extends PlatformPlugin {
...
+  static const String _kEventName = 'integration_test.VmServiceProxyGoldenFileComparator';
+  static const String _kExtension = 'ext.$_kEventName';
+
+  Future<void> _listenToVmServiceForGoldens({required Uri uri, required String testPath}) async {
+    final Uri goldensBaseUri = Uri.parse(testPath);
+    final FlutterVmService vmService = await connectToVmService(uri, logger: logger);
+    final IsolateRef testAppIsolate = await vmService.findExtensionIsolate(_kExtension);
+    await vmService.service.streamListen(_kEventName);
+    vmService.service.onEvent(_kEventName).listen((Event e) async {
+      if (e.extensionKind != 'compare' && e.extensionKind != 'update') {
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart:534]: nit: `!['compare', 'update'].contains(e.extensionKind)` might read a little better.

File: packages/integration_test/lib/src/vm_service_golden_client.dart [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+/// }
+/// ```
+///
+/// or:
+///
+/// ```dart
+/// <String, Object?>{
+///   'result': true /* or possibly false, in the case of 'compare' calls */
+/// }
+/// ```
```
Comment (yjbanov) [inline packages/integration_test/lib/src/vm_service_golden_client.dart:63]: Are the details of the message/response public API? Not sure why these need to be under `///`. Maybe make them `//`?

File: packages/integration_test/lib/src/vm_service_golden_client.dart [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  ///
+  /// @nodoc
+  @visibleForTesting
+  VmServiceProxyGoldenFileComparator.forTesting(this._postEvent);
+
+  static bool get _isRunningOnHost {
+    if (kIsWeb) {
+      return false;
+    }
+    return !io.Platform.isAndroid && !io.Platform.isIOS;
```
Comment (yjbanov) [inline packages/integration_test/lib/src/vm_service_golden_client.dart:86]: Are these the only platforms that separate device from host?

File: packages/integration_test/lib/src/vm_service_golden_client.dart [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+  }
+
+  static void _assertNotRunningOnFuchsia() {
+    if (!kIsWeb && io.Platform.isFuchsia) {
+      throw UnsupportedError('Fuchsia is not supported');
```
Comment (yjbanov) [inline packages/integration_test/lib/src/vm_service_golden_client.dart:91]: Consider a more specific error message, e.g. "integration_test golden testing does not support Fuchsia".

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 1 above)
Comment (matanlurey) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Done.

File: packages/integration_test/lib/src/vm_service_golden_client.dart (hunk 2 above)
Comment (matanlurey) [reply packages/integration_test/lib/src/vm_service_golden_client.dart:63]: It's not clear to me what is/isn't public API here - for example, another package (flutter_tools) needs to fulfill this API for the feature to work.

I am going to keep it as-is for now.

File: packages/integration_test/lib/src/vm_service_golden_client.dart (hunk 3 above)
Comment (matanlurey) [reply packages/integration_test/lib/src/vm_service_golden_client.dart:86]: Currently, yes.

File: packages/integration_test/lib/src/vm_service_golden_client.dart (hunk 4 above)
Comment (matanlurey) [reply packages/integration_test/lib/src/vm_service_golden_client.dart]: Done.

**Pull request merged**